internal/rootcoord/import_manager.go

```diff
@@ -462,181 +596,193 @@ func (m *importManager) getTaskState(tID int64) *milvuspb.GetImportStateResponse
 		defer m.pendingLock.Unlock()
+		index := 0
 		for _, t := range m.pendingTasks {
 			if taskExpired(t) {
-				// Mark this expired task as failed.
-				log.Info("a pending task has expired", zap.Int64("task ID", t.GetId()))
-				t.State.StateCode = commonpb.ImportState_ImportFailed
-				t.State.ErrorMessage = taskExpiredMsgPrefix +
-					(time.Duration(Params.RootCoordCfg.ImportTaskExpiration*1000) * time.Millisecond).String()
-				log.Info("releasing seg ref locks on expired import task",
-					zap.Int64s("segment IDs", t.GetState().GetSegments()))
-				err := retry.Do(m.ctx, func() error {
-					return releaseLockFunc(m.ctx, t.GetId(), t.GetState().GetSegments())
-				}, retry.Attempts(100))
-				if err != nil {
-					log.Error("failed to release lock, about to panic!")
-					panic(err)
+				taskID := t.GetId()
+				m.pendingLock.Unlock()
+				if err := m.setImportTaskState(taskID, commonpb.ImportState_ImportFailed); err != nil {
+					log.Error("failed to set import task state",
+						zap.Int64("task ID", taskID),
+						zap.Any("target state", commonpb.ImportState_ImportFailed))
 				}
-				m.updateImportTaskStore(t)
+				m.pendingLock.Lock()
+				log.Info("a pending task has expired", zap.Int64("task ID", t.GetId()))
+			} else {
+				// Only keep non-expired tasks in memory.
+				m.pendingTasks[index] = t
+				index++
 			}
 		}
+		// To prevent memory leak.
+		for i := index; i < len(m.pendingTasks); i++ {
+			m.pendingTasks[i] = nil
+		}
+		m.pendingTasks = m.pendingTasks[:index]
 	}()
 	// Expire old working tasks.
 	func() {
 		m.workingLock.Lock()
 		defer m.workingLock.Unlock()
 		for _, v := range m.workingTasks {
 			if taskExpired(v) {
-				// Mark this expired task as failed.
 				log.Info("a working task has expired", zap.Int64("task ID", v.GetId()))
-				v.State.StateCode = commonpb.ImportState_ImportFailed
-				v.State.ErrorMessage = taskExpiredMsgPrefix +
-					(time.Duration(Params.RootCoordCfg.ImportTaskExpiration*1000) * time.Millisecond).String()
-				log.Info("releasing seg ref locks on expired import task",
-					zap.Int64s("segment IDs", v.GetState().GetSegments()))
-				err := retry.Do(m.ctx, func() error {
-					return releaseLockFunc(m.ctx, v.GetId(), v.GetState().GetSegments())
-				}, retry.Attempts(100))
-				if err != nil {
-					log.Error("failed to release lock, about to panic!")
-					panic(err)
+				taskID := v.GetId()
+				m.workingLock.Unlock()
+				if err := m.setImportTaskState(taskID, commonpb.ImportState_ImportFailed); err != nil {
+					log.Error("failed to set import task state",
+						zap.Int64("task ID", taskID),
+						zap.Any("target state", commonpb.ImportState_ImportFailed))
 				}
-				m.updateImportTaskStore(v)
+				m.workingLock.Lock()
+				// Remove this task from memory.
+				delete(m.workingTasks, v.GetId())
 			}
 		}
 	}()
 }
 
+// expireOldTasksFromEtcd removes tasks from Etcd that are over `ImportTaskRetention` seconds old.
+func (m *importManager) expireOldTasksFromEtcd() {
```

Comment:
We should keep all tasks in memory.
If rootcoord crashed we reloaded data from etcd.
So we don't need to expireTasks from etcd && memory

Reply:
Discussed offline.

This is from the old design which is not elegant.